Added an empty pass boilerplate as a tutorial (#872)
PennyLaneAI/catalyst: doc/dev/transforms.rst

same representation.

Writing and running your first Catalyst pass
============================================

.. note::

    If you are encoutering issues, or would like to quickly try out the hello world pass described in this
    section, you can have a look at or cherry-pick this commit which includes all changes described

>     in this section: https://github.com/PennyLaneAI/catalyst/commit/ba7b3438667963b307c07440acd6d7082f1960f3

Reviewer: I would maybe put the note at the end.

> If this is your first time writing MLIR or LLVM passes, the boilerplate can be quite overwhelming. 

Reviewer: ```suggestion
If this is your first time writing a MLIR pass, the boilerplate can be quite overwhelming. 
```
I don't think we provide any doc about LLVM passes

Let's first set up the various boilerplate items required to register and run a new pass. 

We'll create an empty pass in the ``Catalyst`` dialect that just prints out hello world to stdout.

> Note that the ``mlir/include`` (and ``mlir/lib``) directories consists of all the available dialects, so if you want to write a new pass in another dialect, it should be added to the subdirectory of that dialect. 

Reviewer: ```suggestion
Note that the ``mlir/include`` and ``mlir/lib`` directories consists of all the available dialects, so if you want to write a new pass in another dialect, it should be added to the subdirectory of that dialect. 
```

> The first thing to do is to create the pass object in the tablegen ``mlir/include/Catalyst/Transforms/Passes.td``:

Reviewer: Can you add a reference to what is tablegen?

Author: done

Tablegen is designed such that we don't have to write all that boilerplate ourselves. 

Now we write the pass itself. Create a new file ``mlir/lib/Catalyst/Transforms/MyHelloWorldPass.cpp`` with the following content:

.. code-block:: cpp

    #define DEBUG_TYPE "myhelloworld"

    #include "Catalyst/IR/CatalystDialect.h"
    #include "mlir/Pass/Pass.h"
    #include "llvm/Support/Debug.h"

    using namespace llvm;
    using namespace mlir;
    using namespace catalyst;

    namespace catalyst {
    #define GEN_PASS_DEF_MYHELLOWORLDPASS
    #define GEN_PASS_DECL_MYHELLOWORLDPASS
    #include "Catalyst/Transforms/Passes.h.inc"

    struct MyHelloWorldPass : public impl::MyHelloWorldPassBase<MyHelloWorldPass> {
        using impl::MyHelloWorldPassBase<MyHelloWorldPass>::MyHelloWorldPassBase;

        void runOnOperation() override { llvm::errs() << "Hello world!\n"; }
    };

    std::unique_ptr<Pass> createMyHelloWorldPass() { return std::make_unique<MyHelloWorldPass>(); }

    } // namespace catalyst

We make the pass object ``MyHelloWorldPass``, which inherits from the base class ``MyHelloWorldPassBase`` that tablegen will build in the namespace ``impl``. 

> The function that determines what your pass actually does is the ``void runOnOperation()``. Here all the pass does is print out ``"Hello world!\n"``. 

Reviewer: ```suggestion
The function that determines what your pass actually does is the ``void runOnOperation()``. Currently, all this pass does is print out ``"Hello world!\n"``. 
```

The other option is the ``errs()`` used here, which will print no matter what.)

This new C++ file needs to be added to the ``mlir/lib/Catalyst/Transforms/CMakeLists.txt`` file (or the CMakeLists.txt of whichever directory that has your new pass file): 

.. code-block::

    file(GLOB SRC
        ...
        MyHelloWorldPass.cpp
    )

After writing the pass, we need to register it in a few places. In ``mlir/include/Catalyst/Transforms/Passes.h``, add the method 

.. code-block:: cpp

    namespace catalyst {
        ...
        std::unique_ptr<mlir::Pass> createMyHelloWorldPass();
        ...
    }

And in ``mlir/lib/Catalyst/Transforms/RegisterAllPasses.cpp``, register the pass via 

.. code-block:: cpp

    void catalyst::registerAllCatalystPasses()
    {
        ...
        mlir::registerPass(catalyst::createMyHelloWorldPass);
        ...
    }

Note that this addition in ``RegisterAllPasses.cpp`` needs to happen in the ``lib/Catalyst/Transforms`` directory, regardless of which dialect your pass belongs to.

> Now that we have written our shiny new pass, we can build it by going back to the top-level ``catalyst`` directory and 

Reviewer: Minor: mention the that it is a command line instruction

.. code-block::

    make dialects

> The tool to run passes is built as ``mlir/build/bin/quantum-opt``. 

Reviewer: ```suggestion
The tool to run passes is the executable `quantum-opt` at the location ``mlir/build/bin/quantum-opt``. 
```

> Since this is an executable, it needs to be invoked as ``./quantum-opt`` instead of just plain ``quantum-opt`` (if you are in the ``mlir/build/bin`` directory; otherwise supply the full path).

Reviewer: Maybe we can add that in the .zshrc we can add the line:

```
export PATH=.../mlir/build/bin:$PATH
```
where ... is the the Catalyst directory location.

Author: done

same representation.

Writing and running your first Catalyst pass
============================================

If this is your first time writing MLIR or LLVM passes, the boilerplate can be quite overwhelming. 
Let's first set up the various boilerplate items required to register and run a new pass. 

We'll create an empty pass in the ``Catalyst`` dialect that just prints out hello world to stdout. The full codelisting of the pass already exists in Catalyst as ``MyHelloWorldPass`` in the directories ``mlir/include/Catalyst/Transforms`` and ``mlir/lib/Catalyst/Transforms``, and we will go over it here. Note that the ``mlir/include`` (and ``mlir/lib``) directories consists of all the available dialects, so if you want to write a new pass in another dialect, it should be added to the subdirectory of that dialect. 

The first thing to do is to create the pass object in the tablegen ``mlir/include/Catalyst/Transforms/Passes.td``:

.. code-block::

    def MyHelloWorldPass : Pass<"my-hello-world"> {
        let summary = "An empty pass boilerplate that prints out hello world.";

        let constructor = "catalyst::createMyHelloWorldPass()";
    }

> When the dialect is built, this tablegen def will be built to a C++ file ``Catalyst/Transforms/Passes.h.inc``, containing the newly defined object called ``MyHelloWorldPass``, alongside the various necessary boilerplate methods in the MLIR infrastructure. Tablegen is designed such that we don't have to write all that boilerplate ourselves. 

Reviewer 2: ```suggestion
When the dialect is built, this tablegen def will be built to a C++ file ``mlir/build/include/Catalyst/Transforms/Passes.h.inc``, containing the newly defined object called ``MyHelloWorldPassBase``, alongside the various necessary boilerplate methods in the MLIR infrastructure. Tablegen is designed such that we don't have to write all that boilerplate ourselves. 
```

perhaps clear to most but it took me a while to find them as there is also just `mlir/include/Catalyst/Transforms/`

Author: This is valuable to add!

Now we write the pass itself. Create a new file ``mlir/lib/Catalyst/Transforms/MyHelloWorldPass.cpp`` with the following content:

.. code-block:: cpp

    #define DEBUG_TYPE "myhelloworld"

    #include "Catalyst/IR/CatalystDialect.h"
    #include "mlir/Pass/Pass.h"
    #include "llvm/Support/Debug.h"

    using namespace llvm;
    using namespace mlir;
    using namespace catalyst;

    namespace catalyst {
    #define GEN_PASS_DEF_MYHELLOWORLDPASS
    #include "Catalyst/Transforms/Passes.h.inc"

>     struct MyHelloWorldPass : public impl::MyHelloWorldPassBase<MyHelloWorldPass> {

Author: The base class does come from the tablegen. I will add this explanation.

>         using impl::MyHelloWorldPassBase<MyHelloWorldPass>::MyHelloWorldPassBase;

Author: yes, `impl` is the namespace where is base class is built in. I will add this in the docs.

        void runOnOperation() override { llvm::errs() << "Hello world!\n"; }
    };

    std::unique_ptr<Pass> createMyHelloWorldPass() { return std::make_unique<MyHelloWorldPass>(); }

    } // namespace catalyst

> The function that determines what your pass actually does is the ``void runOnOperation()``. Here all the pass does is print out ``"Hello world!\n"``. 

Author: llvm has its own printing infrastructure. I have added this in the docs.

This new C++ file needs to be added to the ``mlir/lib/Catalyst/Transforms/CMakeLists.txt`` file (or the CMakeLists.txt of whichever directory that has your new pass file): 

.. code-block::

    file(GLOB SRC
        ...
        MyHelloWorldPass.cpp
    )

After writing the pass, we need to register it in a few places. In ``mlir/include/Catalyst/Transforms/Passes.h``, add the method 

.. code-block:: cpp

    namespace catalyst {
        ...
        std::unique_ptr<mlir::Pass> createMyHelloWorldPass();
        ...
    }

And in ``mlir/lib/Catalyst/Transforms/RegisterAllPasses.cpp``, register the pass via 

.. code-block:: cpp

    void catalyst::registerAllCatalystPasses()
    {
        ...
        mlir::registerPass(catalyst::createMyHelloWorldPass);
        ...
    }

Now that we have written our shiny new pass, we can build it by going back to the top-level ``catalyst`` directory and 

.. code-block::

    make dialects

The tool to run passes is built as ``mlir/build/bin/quantum-opt``.

> We can inspect by all the available passes by running ``quantum-opt --help``:

Reviewer 2: I need to either navigate to `mlir/build/bin/` and then run `./quantum-opt --help` or directly run `./mlir/build/bin/quantum-opt --help`, otherwise it doesnt work. Is that macOS specific and clear for everyone else? Not sure I would have gotten that if @dime10 didnt show me how to use the tool somewhere else

Author: This is how to run executables for any linux-style terminals, including the one on mac. It's the same issue where `a.out` will not run a compiled C program but `./a.out` will. I will add this in the docs as well

.. code-block::

    OVERVIEW: Quantum optimizer driver
    ...
    USAGE: quantum-opt [options] <input file>

    OPTIONS:
        ...
        --my-hello-world                   -   An empty pass boilerplate that prints out hello world.

Here the displayed ``--help`` message will be the ``summary`` we wrote in the tablegen file. The command line option to run our new pass is the template string in the def line in the tablegen file. 

To run the pass, simply do 

.. code-block::

>     quantum-opt -my-hello-world input.mlir

Reviewer 2: ```suggestion
    ./mlir/build/bin/quantum-opt -my-hello-world input.mlir
```

What is the `input.mlir` file expected to contain? I guess you can in principle just run the pass with an empty file? Then I get

```
% ./mlir/build/bin/quantum-opt --my-hello-world hackathon/empty.mlir
Hello world!
module {
}
```

Author: Yes, any input mlir file will do

same representation.

Writing and running your first Catalyst pass
============================================

If this is your first time writing MLIR or LLVM passes, the boilerplate can be quite overwhelming. 
Let's first set up the various boilerplate items required to register and run a new pass. 

> We'll create an empty pass in the ``Catalyst`` dialect that just prints out hello world to stdout. The full codelisting of the pass already exists in Catalyst as ``MyHelloWorldPass`` in the directories ``mlir/include/Catalyst/Transforms`` and ``mlir/lib/Catalyst/Transforms``, and we will go over it here. Note that the ``mlir/include`` (and ``mlir/lib``) directories consists of all the available dialects, so if you want to write a new pass in another dialect, it should be added to the subdirectory of that dialect. 

Author: done

The first thing to do is to create the pass object in the tablegen ``mlir/include/Catalyst/Transforms/Passes.td``:

.. code-block::

    def MyHelloWorldPass : Pass<"my-hello-world"> {
        let summary = "An empty pass boilerplate that prints out hello world.";

        let constructor = "catalyst::createMyHelloWorldPass()";
    }

When the dialect is built, this tablegen def will be built to a C++ file ``mlir/build/include/Catalyst/Transforms/Passes.h.inc``, containing the newly defined object called ``MyHelloWorldPassBase``, alongside the various necessary boilerplate methods in the MLIR infrastructure. Tablegen is designed such that we don't have to write all that boilerplate ourselves. 

Now we write the pass itself. Create a new file ``mlir/lib/Catalyst/Transforms/MyHelloWorldPass.cpp`` with the following content:

.. code-block:: cpp

    #define DEBUG_TYPE "myhelloworld"

    #include "Catalyst/IR/CatalystDialect.h"
    #include "mlir/Pass/Pass.h"
    #include "llvm/Support/Debug.h"

    using namespace llvm;
    using namespace mlir;
    using namespace catalyst;

    namespace catalyst {

>     #define GEN_PASS_DEF_MYHELLOWORLDPASS

Reviewer: Can you add decl as well? Sometimes it is useful

    #include "Catalyst/Transforms/Passes.h.inc"

    struct MyHelloWorldPass : public impl::MyHelloWorldPassBase<MyHelloWorldPass> {
        using impl::MyHelloWorldPassBase<MyHelloWorldPass>::MyHelloWorldPassBase;

        void runOnOperation() override { llvm::errs() << "Hello world!\n"; }
    };

    std::unique_ptr<Pass> createMyHelloWorldPass() { return std::make_unique<MyHelloWorldPass>(); }

    } // namespace catalyst

We make the pass object ``MyHelloWorldPass``, which inherits from the base class ``MyHelloWorldPassBase`` that tablegen will build in the namespace ``impl``. The function that determines what your pass actually does is the ``void runOnOperation()``. Here all the pass does is print out ``"Hello world!\n"``. 

(A sidenote on printing messages in MLIR: there are two major printing options in LLVM. The `more standard one <https://llvm.org/docs/ProgrammersManual.html#the-llvm-debug-macro-and-debug-option>`_ is ``dbgs()``, which only prints when a debug flag is set. The other option is the ``errs()`` used here, which will print no matter what.)

> 

Reviewer: Can you add a small paragraph about patterns? Even if not necessary for that example, maybe just mention that it exists and points to external docs.

Author: Considering the immediate next section is about pattern rewrites (the `MatchAndRewrite` and the `ApplyAndFoldGreedily`), I don't think this boilerplate section needs to mention it.

  .. code-block:: cpp

>       rewriter.updateRootInPlace(op, [&] {

Reviewer: I think it is necessary when modifying op inplace

Author: Ahh, they changed the name https://github.com/llvm/llvm-project/pull/78260
